### PR TITLE
rest: improve RestClient API

### PR DESCRIPTION
**Pull request by rjeczalik**
This CL cleans up RestClient API, fixing couple of incosistencies and
shortcommings, like:

  - when (*RestClient).Options is changed by the user, changes are
    not reflected in *Auth
  - all the tests were modyfing global client instance, not it's more
    COW like
  - ~~RestClient and ClientOptions are fully zero-value friendly~~
  - ClientOptions properly manages TLS / NoTLS transitions

RestClient would require more work, like support for token auth
(it's using BasicAuth for requests), proper TLS/NoTLS testing. This is
what I've seen from the code itself, possibly there's more if compared
to SPEC.md.

~~Changes in Go API:~~

- ~~added function for initializing RestClient from key:~~

  ```
  client, err := ably.NewRestClient("xxx:xxx")
  ```

- ~~RestClient can be initialized from ClientOptions for more control
  over behaviour: (added correct handling of zero-value struct)~~

  ```
  opts := &ably.ClientOptions{Token: "xxx", Secret: "xxx"}
  client := &alby.RestClient{Options: opts}
  ```

- ~~ClientOptions has no `Key` nor `NoTLS` fields, as it's not possible to
  keep them in sync with dependant fields like Token, Secret for Key or
  RestEndpoint, RealtimeEndpoint for `NoTLS`, e.g.~~

  ```go
  /* old API */
  client := ably.NewRestClient(&ably.Options{Key: "xxx:xxx"})
  // ...
  client.Options.Key = "yyy:yyy"
  // all calls to REST API are going to fail, as Token and Secret fields
  // are still "xxx"

  /* new API */
  client, err := ably.NewRestClient("xxx:xxx")
  // ...
  err := client.Options.SetKey("yyy:yyy")
  // correctly handled here
  ```

~~The same applies to NoTLS, where flipping its value should rewrite URLs
of RestEndpoint and RealtimeEndpoint. With new API it's done with:~~

  ```go
  client.SetTLS(false)
  ```

/cc @kouno Please take a look.

BTW I have trello dashboard for ably-go, as I needed to store somewhere my thoughts while working with current library: https://trello.com/b/pHESio8l/ably-go

### CODE REVIEW

File: ably/auth_test.go [hunk 1]
```diff
@@ -10,11 +10,11 @@ import (
 var _ = Describe("Auth", func() {
 	Describe("RequestToken", func() {
 		It("gets a token from the API", func() {
-			req := client.Auth.CreateTokenRequest()
+			req := client.Auth().CreateTokenRequest()
```
Comment (mattheworiordan) [inline ably/auth_test.go]: I'm interested to see that Auth is not a property / attribute, but instead a function.  Why is that? That is inconsistent with other client libraries at present because `Auth` is automatically created for every `Client`.

File: ably/options.go [hunk 2]
```diff
@@ -14,36 +15,72 @@ const (
...
+)
+
+func (opts *ClientOptions) SetTLS(enabled bool) {
+	if enabled {
+		opts.RestEndpoint = restTLS.Replace(opts.RestEndpoint)
+		opts.RealtimeEndpoint = realtimeTLS.Replace(opts.RealtimeEndpoint)
+	} else {
+		opts.RestEndpoint = restNoTLS.Replace(opts.RestEndpoint)
+		opts.RealtimeEndpoint = realtimeNoTLS.Replace(opts.RealtimeEndpoint)
 	}
```
Comment (mattheworiordan) [inline ably/options.go]: This feels like a rather cumbersome approach to what should be a simple task.  The end point should be constructed based on whether the connection is over TLS or not, along with the domain.  Why are we doing string replacing here and above when we could far more easily return an end point when requested that is constructed?

File: ably/options.go [hunk 3]
```diff
@@ -14,36 +15,72 @@ const (
...
+	opts.Token = s[0]
+	opts.Secret = s[1]
+	return nil
+}
+
+func (opts *ClientOptions) Key() string {
+	if opts.Token == "" || opts.Secret == "" {
+		return ""
+	}
+	return opts.Token + ":" + opts.Secret
```
Comment (mattheworiordan) [inline ably/options.go]: The token and key are two completely separate things?  `Token` is an expiring token that is created by a [requestToken](http://docs.ably.io/rest-api/#requesttoken) request, whereas a Key is made up of a key name and key secret, although mostly users will not be separating the two and simply passing a `key` option to the library when instanced.

File: ably/rest_client.go [hunk 4]
```diff
@@ -92,27 +99,24 @@ func (c *RestClient) Stats(params *PaginateParams) (*PaginatedResource, error) {
...
+	req, err := http.NewRequest("GET", c.options().RestEndpoint+path, nil)
 	if err != nil {
 		return nil, err
 	}
 	req.Header.Set("Accept", "application/json")
-	req.SetBasicAuth(c.Auth.Options.Token, c.Auth.Options.Secret)
-	res, err := c.httpclient().Do(req)
-
+	// TODO(rjeczalik): add support for token auth
+	req.SetBasicAuth(c.options().Token, c.options().Secret)
```
Comment (mattheworiordan) [inline ably/rest_client.go]: This is not a token, it's an ApiKey and ApiSecret

File: ably/testutil/test_app.go [hunk 5]
```diff
@@ -64,24 +73,32 @@ func (t *App) populate(res *http.Response) error {
 		return err
 	}
 
-	t.Options.Key = t.AppKeyId() + ":" + t.AppKeyValue()
+	t.opts.Token = t.AppKeyId()
```
Comment (mattheworiordan) [inline ably/testutil/test_app.go]: KeyName & KeySecret

File: ably/options_test.go [hunk 6]
```diff
@@ -1,61 +1,14 @@
...
-		buffer = gbytes.NewBuffer()
-
-		options = &ably.ClientOptions{
-			Key: "id:secret",
-		}
-
-		options.Prepare()
-	})
-
-	It("parses Key into a set of known parameters", func() {
```
Comment (mattheworiordan) [inline ably/options_test.go]: Why are these tests removed?

File: ably/options.go (hunk 2 above)
Comment (rjeczalik) [inline ably/options.go]: It's just a fix, the REST client was implemented that way the Endpoints are URI instead of hostnames, but API did not provide method to switch between TLS/no TLS; I'll rewrite it to make Endpoints just hostnames.

File: ably/testutil/test_app.go (hunk 5 above)
Comment (rjeczalik) [inline ably/testutil/test_app.go]: opts.Token should be opts.ApiKey and t.AppKeyID should be t.ApiSecret, is it right?

File: ably/rest_client.go (hunk 4 above)
Comment (rjeczalik) [inline ably/rest_client.go]: ack

File: ably/options.go (hunk 3 above)
Comment (rjeczalik) [inline ably/options.go]: the opts.Token should be renamed to opts.ApiKey or opts.KeyName? the same for Secret - ApiSecret or KeySecret?

File: ably/options_test.go (hunk 6 above)
Comment (rjeczalik) [inline ably/options_test.go]: Previous version: setting wrong key did not fail but only logged warning - the removed tests checked whether it there was text logged
New version: setting wrong key fails with an error

Ok, I'll add tests for checking whether setting invalid key gives an error.

File: ably/auth_test.go (hunk 1 above)
Comment (rjeczalik) [inline ably/auth_test.go]: Done, changed back to field.

File: ably/options.go (hunk 2 above)
Comment (rjeczalik) [inline ably/options.go]: Done, RestEndpoint and RealtimeEndpoint are hostnames now and URLs are constructed internally based on NoTLS property (must be inverted as the only default we can have in Go for bool is false, thus NoTLS).

File: ably/options.go (hunk 3 above)
Comment (rjeczalik) [inline ably/options.go]: Done, Token renamed to KeyName and Secret to KeySecret (was confused a bit by ApiName / ApiSecret suggestion).

File: ably/rest_client.go (hunk 4 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Done, renamed to KeyName / KeySecret.

File: ably/testutil/test_app.go (hunk 5 above)
Comment (rjeczalik) [inline ably/testutil/test_app.go]: Done, renamed to KeyName / KeySecret.

File: ably/options_test.go (hunk 6 above)
Comment (rjeczalik) [inline ably/options_test.go]: Done, invalid key test-case now covered.

**Comment by rjeczalik** [comment]
@mattheworiordan @kouno Please take a look.

File: ably/auth_test.go (hunk 1 above)
Comment (kouno) [inline ably/auth_test.go]: Hum. I thought this was kind of in-line with the change needed for `Presence`. The only way for a property to be immutable outside the package you have to use accessors.

We discussed this for `Presence` to be `Presence()` instead, so if we enforce such a behaviour elsewhere, I would advise to be consistent with this one too.

File: README.md [hunk 7]
```diff
@@ -31,8 +31,13 @@ TODO
...
 
 ```go
-client := ably.NewRestClient(ably.ClientOptions{ApiKey: "xxxx"})
-channel := client.Channel('test')
+options, err := ably.NewClientOptions("xxx:xxx")
+if err != nil {
+	panic(err)
+}
+
+clien := ably.NewRestClient(options)
```
Comment (kouno) [inline README.md]: Spelling mistake :).

File: ably/options_test.go [hunk 8]
```diff
@@ -1,58 +1,33 @@
...
-		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(opts.KeyName).To(Equal("name"))
+		Expect(opts.KeySecret).To(Equal("secret"))
+	})
 
-		AfterEach(func() {
-			ably.Log.Logger = old
-		})
+	It("when Key is invalid it returns an error", func() {
```
Comment (kouno) [inline ably/options_test.go]: This still reads as a `Context` not a `It`. A context sets some configuration (in this case the fact that we pass an invalid key), and then you write the expectation in the It block.

File: ably/rest_client.go [hunk 9]
```diff
@@ -27,36 +29,30 @@ func query(fn func(string, interface{}) (*http.Response, error)) QueryFunc {
...
-	RestEndpoint string
-	Protocol     string
+	Options *ClientOptions
+	Auth    *Auth
 
-	HTTPClient *http.Client
-
-	channels map[string]*RestChannel
-	chanMtx  sync.Mutex
+	mtx   sync.Mutex // protects chans
```
Comment (kouno) [inline ably/rest_client.go]: That looks weird. Renaming a variable which obfuscates its purpose and forcing us to do a look up to see that it `// protects chans`?

File: ably/rest_client.go [hunk 10]
```diff
@@ -92,27 +86,24 @@ func (c *RestClient) Stats(params *PaginateParams) (*PaginatedResource, error) {
...
-	req, err := http.NewRequest("GET", c.RestEndpoint+path, nil)
+	req, err := http.NewRequest("GET", c.options().restURL()+path, nil)
 	if err != nil {
 		return nil, err
 	}
 	req.Header.Set("Accept", "application/json")
-	req.SetBasicAuth(c.Auth.Options.Token, c.Auth.Options.Secret)
-	res, err := c.httpclient().Do(req)
-
+	// TODO(rjeczalik): add support for token auth
```
Comment (kouno) [inline ably/rest_client.go]: Remove this and add an issue in the repo :wink:.

File: ably/rest_client_test.go [hunk 11]
```diff
@@ -21,38 +21,39 @@ import (
...
 		}
 	)
 
-	Context("with a failing request", func() {
+	Context("with a failing req", func() {
 		var (
-			statusCode int
-			request    *http.Request
+			req    *http.Request
+			cli    *ably.RestClient
```
Comment (kouno) [inline ably/rest_client_test.go]: I don't really understand why you want to truncate all variable names like that. `server`, `client` and `request` were acceptable variable names.

File: ably/rest_client_test.go [hunk 12]
```diff
@@ -21,38 +21,39 @@ import (
...
-						return url.Parse(server.URL)
-					},
+					Proxy: func(req *http.Request) (*url.URL, error) { return url.Parse(srv.URL) },
 				},
 			}
 		}
 	)
 
-	Context("with a failing request", func() {
+	Context("with a failing req", func() {
```
Comment (kouno) [inline ably/rest_client_test.go]: This is supposed to be read as normal english :smile:. You are not supposed to copy variable names in there.

File: ably/rest_http_error.go [hunk 13]
```diff
@@ -15,6 +15,7 @@ type RestHttpError struct {
 func NewRestHttpError(response *http.Response, message string) *RestHttpError {
 	err := &RestHttpError{Response: response, Msg: message}
 	err.readBody()
+	err.Response = nil
```
Comment (kouno) [inline ably/rest_http_error.go]: If you unset this, maybe `Response` should become private?

File: ably/testutil/test_app.go [hunk 14]
```diff
@@ -64,24 +80,32 @@ func (t *App) populate(res *http.Response) error {
...
 
-	t.Options.Key = t.AppKeyId() + ":" + t.AppKeyValue()
+	t.opts.KeyName = t.KeyName()
+	t.opts.KeySecret = t.KeySecret()
 
 	return nil
 }
 
+func (t *App) httpclient() *http.Client {
+	if t.Options != nil && t.opts.HTTPClient != nil {
```
Comment (kouno) [inline ably/testutil/test_app.go]: `t.Options`? Shouldn't that be `t.opts != nil`?

File: ably/export_test.go [hunk 15]
```diff
@@ -1,5 +1,13 @@
...
 	return p.buildPath(base, rel)
 }
+
+func (opts *ClientOptions) RestURL() string {
+	return opts.restURL()
+}
+
+func (opts *ClientOptions) RealtimeURL() string {
+	return opts.RealtimeURL()
+}
```
Comment (kouno) [inline ably/export_test.go]: On a side note, this file has been puzzling me for a bit. Isn't this going to be available outside the package anyway?

File: ably/rest_client_test.go (hunk 11 above)
Comment (rjeczalik) [inline ably/rest_client_test.go]: http://doc.cat-v.org/bell_labs/pikestyle "Variable names"

- `req` is common way of naming `http.Request` variables (https://golang.org/pkg/net/http/)
- `statusCode` the Code part is redundant as the type says its a code (int) and not a status text (string); acutally the variable is removed, but if it'd stayed it'd be names just `status`
- `cli` I agree is artificial, however it conflicts with global `client`, so for readability I named it differently (imo better than client2); suggestions welcome on this one :)

File: ably/rest_http_error.go (hunk 13 above)
Comment (rjeczalik) [inline ably/rest_http_error.go]: It should, I need to rewrite RestHttpError either way as it's a bit useless in current form. Can I do this in separate PR?

File: ably/export_test.go (hunk 15 above)
Comment (rjeczalik) [inline ably/export_test.go]: This is defined only for tests - if tests are external (the ones with _test suffix) and for some reason need to access unexported symbols the pattern is to define them exported for tests, usually in `export_test.go` file. As I removed a number of exported functions from the API I needed to export them for tests. I aim to not use unexported functions in rewritten tests when possible.

File: ably/options_test.go (hunk 8 above)
Comment (rjeczalik) [inline ably/options_test.go]: > A context sets some configuration (in this case the fact that we pass an invalid key)

This is not configuration, this is actual test - setting an invalid key. But I'll wrap it into Context for the Ginkgo sake ;)

File: ably/rest_client.go (hunk 9 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Pretty common, you aim for short variables. In stdlib even structs that have multiple mutex use short mutex names:

```
$ grep -irnH 'mu sync' . | grep protects
./net/http/triv.go:33:	mu sync.Mutex // protects n
./net/singleflight.go:28:	mu sync.Mutex       // protects m
./syscall/srpc_nacl.go:23:	outMu sync.Mutex // protects writing to connection
```

or https://golang.org/src/net/rpc/server.go?s=5638:5874#L174. But I'd rename it as I have no strong opinion about it.

File: ably/auth_test.go (hunk 1 above)
Comment (rjeczalik) [inline ably/auth_test.go]: > The only way for a property to be immutable outside the package is to use accessors.

This is not necessarily true, the same problem is in Java - even though you return a variable with accessor without defensive copying it's still can be modified. So we gain nothing with using accessors. However I very like using accessor vs exported fields (exported fields should be used only for model, not control imho, sorry for offtop), as then I can make them being constructed lazily, which allows for using zero-value `RestClient`, which is preferable in Go.

Comming back to `Presence`, the `RestPresence` does not have any fields exported, so the user can't modify it either way :)

File: ably/rest_client.go (hunk 10 above)
Comment (rjeczalik) [inline ably/rest_client.go]: This was actually the first thing I've noticed - I've collected a fair amount things that need either to be fixed, missing or polished a bit, I keep them in my trello board as I didn't want to create noise in github issues - https://trello.com/b/pHESio8l/ably-go. Or you feel you want to give feedback on those and I should move them to github isseus?

File: ably/rest_client.go (hunk 9 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Done, renamed to chansMtx.

File: ably/options_test.go (hunk 8 above)
Comment (rjeczalik) [inline ably/options_test.go]: Done, splitted into Context/BeforeEach and It scopes.

File: README.md (hunk 7 above)
Comment (rjeczalik) [inline README.md]: Done.

File: ably/rest_client.go (hunk 10 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Done, I've removed those TODOs and the one from auth.go about :query_time.

File: ably/rest_client_test.go (hunk 12 above)
Comment (rjeczalik) [inline ably/rest_client_test.go]: Done, this was sloppy sed fail.

File: ably/rest_http_error.go (hunk 13 above)
Comment (rjeczalik) [inline ably/rest_http_error.go]: Done, changed to unexported field.

File: ably/testutil/test_app.go (hunk 14 above)
Comment (rjeczalik) [inline ably/testutil/test_app.go]: Done, thanks.

**Comment by rjeczalik** [comment]
@kouno Another round of fixes, please take a look :)

File: ably/options.go [hunk 16]
```diff
@@ -14,36 +15,86 @@ const (
...
+	if host == "" {
+		host = DefaultOptions.RealtimeEndpoint
+	}
+	if opts.NoTLS {
+		return "ws://" + host + ":80"
+	}
+	return "wss://" + host + ":443"
+}
+
+func (opts *ClientOptions) SetKey(key string) error {
```
Comment (mattheworiordan) [inline ably/options.go]: Sorry, I should have picked up on this earlier, but I'm a little concerned that we're making options mutable.  In Ruby at least, the options are immutable once the client library has been instanced.  This is necessary as if you want a new client with different credentials for example, you'd need to create a new client library.  If I read this correctly, we are making this not only immutable, but in fact exposing public methods making it apparent it is OK to change the options?  Do we mean to do this?

File: README.md [hunk 17]
```diff
@@ -31,8 +31,13 @@ TODO
 All examples assume a client and/or channel has been created as follows:
 
 ```go
-client := ably.NewRestClient(ably.ClientOptions{ApiKey: "xxxx"})
-channel := client.Channel('test')
+options, err := ably.NewClientOptions("xxx:xxx")
```
Comment (mattheworiordan) [inline README.md]: Out of interest, if ClientOptions is effectively just a data type, do you think this style is necessary?  Could we not simply do the err checking in the Client instead and keep ClientOptions as a simple immutable set of options used to instance a Client?

File: ably/options_test.go [hunk 18]
```diff
@@ -1,57 +1,44 @@
...
-		options = &ably.ClientOptions{
-			Key: "id:secret",
-		}
+	Context("when Key is valid", func() {
+		var err error
 
-		options.Prepare()
-	})
+		BeforeEach(func() {
+			err = opts.SetKey("name:secret")
```
Comment (mattheworiordan) [inline ably/options_test.go]: As mentioned above, this is not an expected way to use the options, perhaps this is more idiomatic Go and we can offer this, but I would have expected that one only ever creates the options in one statement such as `ably.ClientOptions { option: "val" }`

File: ably/rest_client.go [hunk 19]
```diff
@@ -27,36 +29,30 @@ func query(fn func(string, interface{}) (*http.Response, error)) QueryFunc {
...
-	Protocol     string
+	Options *ClientOptions
+	Auth    *Auth
 
-	HTTPClient *http.Client
-
-	channels map[string]*RestChannel
-	chanMtx  sync.Mutex
+	chansMtx sync.Mutex
+	chans    map[string]*RestChannel
```
Comment (mattheworiordan) [inline ably/rest_client.go]: Why change of channels to chans which is less descriptive?

File: ably/rest_client.go [hunk 20]
```diff
@@ -66,21 +62,19 @@ func (c *RestClient) Time() (time.Time, error) {
 		return time.Time{}, err
 	}
 	if len(times) != 1 {
-		return time.Time{}, fmt.Errorf("Expected 1 timestamp, got %d", len(times))
+		return time.Time{}, fmt.Errorf("expected 1 timestamp, got %d", len(times))
```
Comment (mattheworiordan) [inline ably/rest_client.go]: Why lower case "expected" whereas below you have "Unexpected" i.e. sentence case.  I'm not clear on why this commit was even made given it creates inconsistencies in case now?

File: ably/rest_client_test.go (hunk 11 above)
Comment (mattheworiordan) [inline ably/rest_client_test.go]: I am with @kouno on this, I don't see the benefit of `srv` of `server` or `req` over `request`.  Whilst perhaps this shorter notation is used in other packages, I would prefer to have our variables descriptive.  `status` is fine, however please note that in our errors we have `status` and `code` where code is our internal error code.

Comment (mattheworiordan) [inline ably/rest_client_test.go]: Why does `client` conflict with a global `client`?  What is the global `client`?

File: ably/rest_client_test.go [hunk 21]
```diff
@@ -85,58 +87,52 @@ var _ = Describe("RestClient", func() {
...
 
-				err := client.Channel("test").Publish("ping", "pong")
+				err := cli.Channel("test").Publish("ping", "pong")
 				Expect(err).NotTo(HaveOccurred())
 			})
 
 			It("encode the body of the message using msgpack", func() {
 				var anyMsgPack []map[string]interface{}
-				err := msgpack.Unmarshal(buffer, &anyMsgPack)
+				err := msgpack.Unmarshal(buf, &anyMsgPack)
```
Comment (mattheworiordan) [inline ably/rest_client_test.go]: As above, `buf` and `cli` instead of `buffer` and `client` IMHO does not make this code more readable.

File: ably/rest_http_error.go [hunk 22]
```diff
@@ -6,15 +6,16 @@ import (
 )
 
 type RestHttpError struct {
-	Msg      string
-	Response *http.Response
-
+	Msg          string
 	ResponseBody string
+
+	resp *http.Response
```
Comment (mattheworiordan) [inline ably/rest_http_error.go]: It feels odd to have an HTTP Error without the status code as part of the type?  Also, note that our responses always follow a generic format in JSON, see https://rest.ably.io/ as an example.  We therefore have an Ably ErrorInfo type, see https://github.com/ably/ably-java/blob/master/src/io/ably/types/ErrorInfo.java.  Should we not follow that format?

File: ably/rest_client_test.go (hunk 11 above)
Comment (rjeczalik) [inline ably/rest_client_test.go]: Global client is a test one which is a global shared resource for the ginkgo spec. This is extremely dangerous, as part of the tests modify (!) the global instance, so while it works now, it is really fragile. For now I need to live with this, however I'm planning to drop this after all tests are rewritten to testing.

I'll renamed those variables to their original long form a bit later.

File: ably/options.go (hunk 16 above)
Comment (rjeczalik) [inline ably/options.go]: The problem the setter is solving is dependant fields: setting key field should set keyName and keySecret to proper values, However if user sets key to different value (e.g. after instantiating client but before making first request or changing the key in some handler code when the old key was wrong) the keyName and keySecret fields have still the old value, which leads to sloppy API. Thus I solved this with setter. If we have better options or we don't care the please give me feedback. 

From the description:

```
/* old API */
client := ably.NewRestClient(&ably.Options{Key: "xxx:xxx"})
// ...
client.Options.Key = "yyy:yyy"
// all calls to REST API are going to fail, as Token and Secret fields
// are still "xxx"

/* new API */
client, err := ably.NewRestClient("xxx:xxx")
// ...
err := client.Options.SetKey("yyy:yyy")
// correctly handled here
```

File: README.md (hunk 17 above)
Comment (rjeczalik) [inline README.md]: Yes, I wanted to do that in the original version, but in order to make that possible, I needed to have meaningful zero-value for ably.RestClient, so the factory method could take the key string intead of options (Go has no method overloading). The other options is to have to factory methods, one NewRestClient(key string) and second NewRestClientOptions(*CientOptions). If you like the latter idea, I can make it.

The factory method for ClientOptions returns error because it's parsing argument, and parsing can fail.

File: ably/options_test.go (hunk 18 above)
Comment (rjeczalik) [inline ably/options_test.go]: Dependant fields problem, if you can tell me how to solve that differently or that we don't care, then I'll change that.

File: ably/rest_client.go (hunk 19 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Already fixed.

File: ably/rest_client.go (hunk 20 above)
Comment (rjeczalik) [inline ably/rest_client.go]: I'll change all text messages to have lower-case names a bit latter. This is how idiomatic Go error messages are.

File: ably/rest_client_test.go (hunk 21 above)
Comment (rjeczalik) [inline ably/rest_client_test.go]: Sure, will change to longer versions.

File: ably/rest_http_error.go (hunk 22 above)
Comment (rjeczalik) [inline ably/rest_http_error.go]: As I already said, I'm going to rewrite RestHttpError in a separate PR, as it's useless in current form. Is that ok?

File: ably/rest_client_test.go (hunk 11 above)
Comment (kouno) [inline ably/rest_client_test.go]: @rjeczalik you have a valid point about the `client`. This was easy to manage at first because there was not much tests and the `client` was never really changed, but with the tests around client options, this will inevitably clash.

File: ably/rest_client.go (hunk 10 above)
Comment (kouno) [inline ably/rest_client.go]: I don't really like Trello in general... But ignoring that, I think that since this is an open source client library, the best is to place these things in our issues and add a label for `features` for transparency.

Comment (rjeczalik) [inline ably/rest_client.go]: Sure, I'll move them later. I was intended only for my personal use, to I keep track of things.

File: ably/rest_http_error.go (hunk 22 above)
Comment (rjeczalik) [inline ably/rest_http_error.go]: > Also, note that our responses always follow a generic format in JSON, see (...)

Yes, I aim to have single `ably.Error` type used for all client errors with proper constants and fields to denote Code, internal Code etc. I'm just not describing the design here in order to not draw attention away from this PR :)

File: ably/rest_client_test.go (hunk 11 above)
Comment (mattheworiordan) [inline ably/rest_client_test.go]: Ok, thanks for clarifying, well I would change the globalClient if that is the problem then as that part feels like the issue.

File: ably/options.go (hunk 16 above)
Comment (mattheworiordan) [inline ably/options.go]: I think the problem here is that we're allowing someone to change the key afterwards at all.  Why are we allowing this?  It is unnecessary and introduces these problems.
Also, like you changed with the "realtimeEndPoint", `keyName` and `keySecret` (not Token & Secret) can be determined at runtime from the full Key.

File: README.md (hunk 17 above)
Comment (mattheworiordan) [inline README.md]: I think the shorthand initialisation with just a key string is causing too many problems and if overloading is not possible, I would simply go for the following instead:

```
ably.NewRestClient(ably.ClientOptions{ Key: "xxxx" })
```

It reduces complexity significantly.

File: ably/options_test.go (hunk 18 above)
Comment (mattheworiordan) [inline ably/options_test.go]: Why not:
```
BeforeEach(func() {
   client, err = ably.NewRestClient(ably.ClientOptions{Key:"name:secret"})
}
````

File: ably/options.go (hunk 16 above)
Comment (kouno) [inline ably/options.go]: > I think the problem here is that we're allowing someone to change the key afterwards at all. Why are we allowing this? It is unnecessary and introduces these problems.

Agreed. Once instantiated, the options shouldn't be modifiable. You have to set the options right before instantiating.

Comment (rjeczalik) [inline ably/options.go]: But then *ClientOptions should not be exported / accessible from `RestClient`, and right now it is; I'd love to unexport it, since once ClientOptions were passed from outside to client, they should also be accessed from outside and not from the `RestClient` - thanks to this change user would have no opportunity of altering this field; however I've seen Java library makes this field public. So the question is - are those ClientOptions needed to be accessbile from RestClient? If not, the it solves all the problems.

Comment (mattheworiordan) [inline ably/options.go]: Yes, perhaps Java has left these options as mutable, but this is not our intended design. @paddybyers do you think we should modify the Java library to make `ClientOptions Rest.options` immutable?  I don't see any valid use case where the options should be changed after the library is created.

Comment (kouno) [inline ably/options.go]: I think we have a misunderstanding here.

This struct should be available to developers, but once you pass the ClientOptions instance in the client, you should not expect that any later change to the instance will be reflected on the Client. For example the following code should not work:

```go
options := ably.NewClientOptions()
options.Key = "something"

client, err = ably.NewRestClient(options)

options.Key = "somethingelse"

// client should still use the previous key.
```

Comment (mattheworiordan) [inline ably/options.go]: Indeed, although I am not sure why we even need to provide the write accessor `options.Key = "something"` when one can simply use `ClientOptions { key: "something" }`

Comment (paddybyers) [inline ably/options.go]: > I don't see any valid use case where the options should be changed after the library is created

No, we should not attempt to support modification of the options and if the language provides a convenient way to ensure they are immutable after having been passed to the `Ably` library then we should do that.

However making them `final` in, say, the java lib then gives you an API that is much more awkward to use because they would then all need to be specified in the `ClientOptions` constructor and I don't think we should do that.

Comment (kouno) [inline ably/options.go]: True. I was just using the current API to make myself clear ;).

Comment (rjeczalik) [inline ably/options.go]: But still RestClient uses its own Options field to e.g. read RestEndpoint. If user changes RestClient.Options.RestEndpoint then there's no way to prevent that other than documenting it that it must be accessed only for reading:

```go
options := ably.NewClientOptions()
options.Key = "something"

client, err := ably.NewRestClient(options)

client.Options.RestEndpoint = "invalid" // this is going to break all the requests made by this client
```

File: README.md (hunk 17 above)
Comment (rjeczalik) [inline README.md]: Do you agree for removing KeyName and KeySecret fields from ClientOptions and have getters instead, which read Key in the runtime?

```go
ops := ably.ClientOptions{Key: "xxx:xxx"}
keyName := opts.KeyName()
keySecret := opts.KeySecret()
```

File: ably/options_test.go (hunk 18 above)
Comment (rjeczalik) [inline ably/options_test.go]: This test-case tests options, not RestClient, I'm not convinced we should mixed responsibilities in the tests. However I'm going to revisit this tests after final modifications to this PR are applied according to other discussions.

File: ably/options.go (hunk 16 above)
Comment (kouno) [inline ably/options.go]: @rjeczalik Well you are not supposed to keep the `Options`. At the moment the library does the right thing for RestEndpoint.

https://github.com/ably/ably-go/blob/master/ably/rest_client.go#L44-L45

Since a string is immutable, modifying `options.RestEndpoint` in my example wouldn't break. It is another story for the http client though...

Comment (mattheworiordan) [inline ably/options.go]: Well we should ideally prevent the options from being editable, so arguably we could make Options not public and expose the options a user may need to access as accessors of the RestClient such as `client.Key()`, but I think you'll need to advise what is possible with Go.

File: README.md (hunk 17 above)
Comment (mattheworiordan) [inline README.md]: Yup

File: ably/options.go (hunk 16 above)
Comment (rjeczalik) [inline ably/options.go]: > Well you are not supposed to keep the Options. At the moment the library does the right thing for RestEndpoint.

Not really, there's no difference, user can change both, either `client.ClientOptions.RestEndpoint` or `client.RestEndpoint`.

Comment (mattheworiordan) [inline ably/options.go]: Ok, well I think we should ideally make Options not a public struct type, and instead provide accessors once the library is instanced so that unless someone intentionally pokes into the library and tries to break things, they will not mistakenly be led to believe they can change attributes.

Comment (rjeczalik) [inline ably/options.go]: Done, Options is not kept within RestClient, Endpoint and Protocol fields are back, they're read-only. Will document that fact with in-line code documentation in separate PR.

File: README.md (hunk 17 above)
Comment (rjeczalik) [inline README.md]: Done, Key field is back with KeyName() and KeySecret().

File: ably/options_test.go (hunk 18 above)
Comment (rjeczalik) [inline ably/options_test.go]: Done, tests for ClientOptions use RestClient.

File: ably/rest_client.go (hunk 20 above)
Comment (rjeczalik) [inline ably/rest_client.go]: Done, error messages start with lower case.

File: ably/rest_client_test.go (hunk 21 above)
Comment (rjeczalik) [inline ably/rest_client_test.go]: Done, long names.

**Comment by rjeczalik** [comment]
@mattheworiordan @kouno: Please take a look, pretty hot PR, I've addressed all naming, options-related issues.

**Comment by rjeczalik** [comment]
@mattheworiordan @kouno: Please take a look, pretty hot PR, I've addressed all naming, options-related issues.

**Comment by kouno** [comment]
Looks ok to me. @mattheworiordan any other comments?

File: ably/options.go [hunk 23]
```diff
@@ -14,36 +15,74 @@ const (
 	ProtocolMsgPack = "msgpack"
 )
 
+var errInvalidKey = errors.New("invalid key format")
+
+var DefaultOptions = &ClientOptions{
+	RestEndpoint:     "rest.ably.io",
+	RealtimeEndpoint: "realtime.ably.io",
```
Comment (mattheworiordan) [inline ably/options.go]: Sorry, should have noticed before, but the option name could be considered a bit misleading as an end point is typically a FQDN i.e. "http://rest.ably.io" whereas what we actually allow is simply a custom host.  See https://github.com/ably/ably-java/blob/master/src/io/ably/types/ClientOptions.java#L61-L67 for the Java implementation.  Personally I prefer Rest & Realtime over the Java implementation, but can we use the name host instead please as this is more accurate and consistent, and we'll clean up the Java version too to match this?

File: ably/options.go [hunk 24]
```diff
@@ -14,36 +15,74 @@ const (
...
 	NoTLS            bool
 
 	HTTPClient *http.Client
 }
 
-func (p *ClientOptions) Prepare() {
-	if p.Key != "" {
-		p.parseKey()
+func (opts *ClientOptions) restURL() string {
+	host := opts.RestEndpoint
```
Comment (mattheworiordan) [inline ably/options.go]: as above, this confirms that `RestEndPoint` is indeed just `RestHost`

File: ably/paginated_resource_test.go [hunk 25]
```diff
@@ -16,7 +16,7 @@ var _ = Describe("PaginatedResource", func() {
 		})
 
 		It("returns a string pointing to the new path based on the given path", func() {
-			newPath, err := ably.BuildPath(&paginatedResource, "/path/to/resource?hello", "./newresource?world")
+			newPath, err := paginatedResource.BuildPath("/path/to/resource?hello", "./newresource?world")
```
Comment (mattheworiordan) [inline ably/paginated_resource_test.go]: This is my fault because in Ruby I named the class `PaginatedResource`, but in Ably-js and Ably-Ruby it's called `PaginatedResult`.  Sorry, entirely my fault I expect.  See https://github.com/ably/ably-ruby/issues/40

File: ably/options.go (hunk 23 above)
Comment (rjeczalik) [inline ably/options.go]: `RestHost` and `RealtimeHost`, correct?

**Comment by mattheworiordan** [comment]
@rjeczalik looks good, only one or two small comments, thanks :+1:

File: ably/options.go (hunk 23 above)
Comment (mattheworiordan) [inline ably/options.go]: Yup, see https://github.com/ably/ably-java/issues/22

**Comment by rjeczalik** [comment]
@mattheworiordan @kouno Please take a look.
